README.md
# `AggAPI.lua`

This UDF will help you with arbitrary aggregations on the Aerospike database. You need to register this lua UDF on the server using your client's API, and then use aggregation API on your chosen Aerospike Client to call it with required parameters.

## Limitations

Aerospike Server supports Lua 5.1, in which all numbers are floats with 51 bits significands (52 with sign bit). This means integers bigger than 2^51 will return invalid values.

## How to setup?

You need to register the provided `aggAPI.lua` file as a UDF in your database. 

#### Using Go to register the module:
Here an example in Go (note that for the sake of conciseness, the errors are not checked in this example):
```go
luaFile, _ := ioutil.ReadFile("aggAPI.lua")
regTask, _ := client.RegisterUDF(nil, luaFile, "aggAPI.lua", aero.LUA)
// wait until UDF is created on the server.
_ <-regTask.OnComplete()
```
For more information: [Go Client register UDF](https://www.aerospike.com/docs/client/go/usage/udf/register.html).

#### Using Java to register the module:
```java
RegisterTask task = client.register(params.policy, "udf/aggAPI.lua", "aggAPI.lua", Language.LUA);
// Alternately register from resource.
task.waitTillComplete();
```
For more information: [Java Client register UDF](https://www.aerospike.com/docs/client/java/usage/udf/register.html).

#### Using AQL to register the module:
```
aql> register module 'aggAPI.lua'
```
For more information: [manging UDF using AQL](https://www.aerospike.com/docs/tools/aql/udf_management.html).

## How does it work?

The Lua [streaming UDF](https://www.aerospike.com/docs/udf/developing_stream_udfs.html) will use the argument you pass to it in its calculations by `eval`ing the arguments, and then using them in its logic to calculate and filter the records mostly on the server-side. 
```json
    "fields": {
        "name":            "name",
        "max(age)":        {"func": "max", "expr": "rec['age'] ~= nil and rec['age']"},
        "count(age)":      {"func": "count", "expr": "( rec['age'] ) ~= nil and 1"},
        "min(age)":        {"func": "min", "expr": "rec['age'] ~= nil and rec['age']"},
        "sum(age*salary)": {"func": "sum", "expr": " (rec['age']  or 0) * (rec['salary'] or 0)"},
    },
    "filter": "rec['age'] ~= nil and rec['age'] >5 ",
    "group_by_fields": {
      "name"
    }
```

A last pass will occur on client-side and the results will return in the following format:
```json
{
  "8de6a795aaf29f2a7dad71c6631a1efc": {
    "count(age)":      3.000000,
    "max(age)":        45.000000,
    "count":           3.000000,
    "sum(age*salary)": 101400,
    "min(age)":        25.000000,
    "name":            "Eva",
  },
  "ed57af7ff6ed54ec8b6b5eec3e2b649a": {
    "count(age)":      1.000000,
    "max(age)":        26.000000,
    "count":           1.000000,
    "sum(age*salary)": 83200,
    "min(age)":        26.000000,
    "name":            "Riley",
  },
}
```
The `key` is a hash used to group the results for reduction. The value is a map of the returned fields. In the map, the key is the alias of the field.

Keep in mind that the values are limited to the size of Lua's value size, which is 51 bits of significant integer values.

## What is the meaning of the values sent to the UDF?

There are 3 different input that need to be sent to the Lua UDF. Not all are required for every command. These values are:

- `"fields"`: Choosing the fields to return - this is the equivalent of the `select` part of the query.
    - Fields which do not require any complex calculation: the map key is the alias, while the map value is the name of the existing bin in the database.  
    Example:
      ```json
      fields": {
        "age": "age",  
        "salary_usd" : "salary"  
      }
      ```
    - Fields which are calculated (apply an aggregate function on): the map key is the aliases, and the value is a map of the function and its calculation.
      Available functions are `count`, `sum`, `min`, `max`.   
             
      Example:
      ```json
      "fields": {
        "sum(salary * 2)": {"func": "sum", "expr": "rec['salary'] * 2"}, 
        "min(salary)":     {"func": "sum", "expr": "rec['salary']"},
        "max(salary * 5)": {"func": "sum", "expr": "(rec['salary'] or 0) * 5"}
      }
      ```
        - `"sum(salary * 2)": {"func": "sum", "expr": "rec['salary'] * 2"}` means a field with the name `sum(salary * 2)` should be calculated from the value of the bin `salary` multiplied by 2. 
        - `"min(salary)":     {"func": "sum", "expr": "rec['salary']"}`: use the value of the bin `salary` to calculate `min(salary)`
        - `"max(salary * 5)": {"func": "sum", "expr": "(rec['salary'] or 0) * 5"}`: use the value of the bin `salary` multiplied by 5 to calculate the max value. If the `salary` bin is `null`, 0 will be used as default value.
  
- `"filter"`: Filter is a lua boolean statement to filter records - this is the equivalent of a `where` in a query.  
  If the value of the `statement` is `true`, the record will be included in the results.
  
  Example:   
   `"filter": "rec['age'] ~= nil and rec['age'] > 25"`
  
- `"group_by_fields"`: List of field aliases to group the records by - this is the equivalent of `group by` in a query.   
Example:
    ```json
    "group_by_fields": [
        "age", "salary_udf"
     ]
    ```

## Example: Building a Query

### How can I calculate a sum?

To calculate the equivalent of the following SQL:

```sql
select sum(salary) from employees
```

We would call the `select_agg_records` function with following argument:

```json
{
  "fields":         {
    "sum(salary)": {"func": "sum" , "expr": "rec['salary'] or 0"},
  },
}
```
Which means use the value of the bin `salary`. If the `salary` bin is `null` or does not exist, 0 will be used as default value.
We name this calculation `sum(salary)`.

### Adding a conditions

To add a filtering condition:

```sql
select sum(salary) from employees where age > 25
```

provide the following arguments:

```json
{
  "fields":         {
    "sum(salary)": {"func": "sum" , "expr": "rec['salary'] or 0"},
  },
  "filter":    "rec['age'] ~= nil and rec['age'] > 25",
}
```

Which means check if the `age` bin exists, and if it does, check if the age is over 25.

### Adding a `Group BY`

To create a group by aggregation, we would need to add the fields both to the `fields` part and to the group by.

```sql
select name, age, sum(salary) from employees where age > 25 group by name, age
```
provide the following arguments:
```json
{
  "fields":         {
    "age":  "age",
    "name": "name",
    "sum(salary)": {"func": "sum" , "expr": "rec['salary'] or 0"},
  },
  "filter":    "rec['age'] ~= nil and rec['age'] > 25",
  "group_by_fields": [
    "age",
    "name",
  ],
}
```
**Please note: the UDF logic does not validate if your provided `group by` arguments are valid for the logic of the equivalent SQL command.**

### Adding `min` and `max` functions

To add other fields to the query like:

```sql
select age, min(salary), max(salary), sum(salary) from employees where age > 25 group by age
```

provide the following arguments:

```json
{
  "fields":         {
    "age": "age",
    "min(salary)": {"func": "min" , "expr": "rec['salary']"},
    "max(salary)": {"func": "max" , "expr": "rec['salary']"},
    "sum(salary)": {"func": "sum" , "expr": "rec['salary']"},
  },
  "filter":    "rec['age'] ~= nil and rec['age'] > 25",
  "group_by_fields": [
    "age",
  ],
}
```

### Can I do more complex statements in the functions and filters?

YES! The execute the equivalent of the following SQL command:

```sql
select age, min(salary), max(salary), sum(salary * 2) from employees where age > 25 group by age
```

provide the following arguments:

```json
{
  "fields":         {
    "age": "age",
    "sum(salary * 2)": {"func": "sum" , "expr": "rec['salary'] * 2"},
    "min(salary)":     {"func": "min" , "expr": "rec['salary']"},
    "max(salary)":     {"func": "max" , "expr": "rec['salary']"},
  },
  "filter":    "rec['age'] ~= nil and rec['age'] > 25",
  "group_by_fields": [
    "age",
  ],
}
```

## How can I calculate average?

The UDF does not have an `avg` function so there is no equivalent for this SQL query:
```sql
select age, avg(salary) from employees group by age
```
Calculating the average needs to be done manually by the client - query `sum` and `count` and calculate the `avg` from them.

Using the UDF to get sum and count:
```json
{
  "fields":         {
    "age": "age",
    "sum(salary)":     {"func": "sum" , "expr": "rec['salary']"},
    "count(*)":        {"func": "count" , "expr": "1"}
  },
  "group_by_fields": [
    "age",
  ],
}
```

## How can I do `DISTINCT` queries?

In case you would want to return the following SQL statement:

```sql
select distinct age from employees
```

since you can rewrite the above query as:

```sql
select age from employees group by age
```

then the parameters sent to the UDF would be:

```json
{
  "fields": {
    "age": "age",
  },
  "group_by_fields": [
    "age",
  ],
}
```

## Code Examples
### Example in Go:
```go
stm := aero.NewStatement(nsName, setName)

functionArgsMap := map[string]interface{}{
  "fields": map[string]interface{}{
    "name":            "name",
    "max(age)":        map[string]string{"func": "max", "expr": "rec['age'] ~= nil and rec['age']"},
    "count(age)":      map[string]string{"func": "count", "expr": "( rec['age'] ) ~= nil and 1"},
    "min(age)":        map[string]string{"func": "min", "expr": "rec['age'] ~= nil and rec['age']"},
    "sum(age*salary)": map[string]string{"func": "sum", "expr": " (rec['age']  or 0) * (rec['salary'] or 0)"},
  },
  "filter": "rec['age'] ~= nil and rec['age'] >5 ",
  "group_by_fields": []string{
    "name",
  },
}

recordset, err := client.QueryAggregate(nil, stm, "aggAPI", "select_agg_records", aero.NewValue(functionArgsMap))
defer recordset.Close()

if err != nil {
  return err
}

for result := range recordset.Results() {
  if result.Err != nil {
    return result.Err
  }

  pp.Println(result.Record.Bins["SUCCESS"])
}
```

### Example in Java:
```java
String stringToParse = String.format("{\n" +
    "  \"fields\":         {\n" +
    "    \"test_id\": \"test_id\",\n" +
    "    \"state\": \"state\",\n" +
    "    \"count(state)\": {\"func\":\"count\", \"expr\": \"rec['state'] ~= nil and 1\"},\n" +
    "  },\n" +
    "  \"filter\":    \"rec['test_id'] ~= nil and rec['test_id'] == %s\",\n" +
    "  \"group_by_fields\": [\n" +
    "    \"test_id\",\n" +
    "    \"state\",\n" +
    "  ],\n" +
    "}", testId.toString());

JSONObject json = new JSONObject(stringToParse);
Map functionArgsMap  = toMap(json);

Statement stmt = new Statement();
stmt.setNamespace(namespace);
stmt.setSetName(set);
stmt.setBinNames(binName);

// Optional filter via Index
stmt.setFilter(Filter.equal(binName, testId));
stmt.setIndexName(indexName);

ResultSet rs = client.queryAggregate(null, stmt,"aggAPI","select_agg_records", Value.get(functionArgsMap));

try {
  if (rs.next()) {
    Object obj = rs.getObject();

    if (obj instanceof Map<?, ?>) {
      Map<String, Map> map = (Map<String, Map>) obj;

        map.values().forEach(res -> {
            console.info(res.toString());

            console.info("res: %s, %s => %s",
                    res.get("test_id"),
                    res.get("state"),
                    res.get("count(*)"));
      }
    }
  }
}

finally {
  rs.close();
}
```


CHANGELOG.md
# Change History

## April 3 2020: v1.0.1

  * **Improvements**

  	- Adds tests against sqlite3 to make sure the results are correct.

  * **Fixes**

    - Fixes a an issue where sometimes more fields than specified would be returned.


## April 1 2020: v1.0.0

  Major Breaking Change.

  * **Changes**

  	- Simplifies the payload definition of the lua function. Read the README.md for more info.

  * **Fixes**

    - Fixes a race condition triggered in the Go client's Lua library.



test/aero_test.go
package main_test

import (
	"io/ioutil"
	"log"
	"path/filepath"

	aero "github.com/aerospike/aerospike-client-go"
)

func aeroClient(host string, port int, user, password, currentPath string) (*aero.Client, error) {
	/****************************************************************************

	Setup Aerospike Client

	****************************************************************************/
	clientPolicy := aero.NewClientPolicy()
	if user != "" {
		clientPolicy.User = user
		clientPolicy.Password = password
	}

	client, err := aero.NewClientWithPolicyAndHost(clientPolicy, aero.NewHost(host, port))
	if err != nil {
		return nil, err
	}

	/****************************************************************************

	Setup Aerospike Lua Path

	****************************************************************************/
	aero.SetLuaPath(filepath.Clean(currentPath) + "/")

	return client, nil
}

func toInt64(v interface{}) int64 {
	switch v := v.(type) {
	case int:
		return int64(v)
	case int64:
		return v
	case float64:
		return int64(v)
	default:
		panic("Not possible")
	}
}

func aeroQuery(client *aero.Client, nsName, setName string, payload map[string]interface{}) ([]map[string]interface{}, error) {
	stm := aero.NewStatement(nsName, setName)

	recordset, err := client.QueryAggregate(nil, stm, "aggAPI", "select_agg_records", aero.NewValue(payload))
	defer recordset.Close()

	if err != nil {
		return nil, err
	}

	res := []map[string]interface{}{}
	for result := range recordset.Results() {
		if result.Err != nil {
			return nil, result.Err
		}

		v := result.Record.Bins["SUCCESS"].(map[interface{}]interface{})
		for _, prec := range v {
			rres := map[string]interface{}{}
			for k, v := range prec.(map[interface{}]interface{}) {
				if i, ok := v.(float64); ok {
					rres[k.(string)] = toInt64(i)
				} else if i, ok := v.(int); ok {
					rres[k.(string)] = toInt64(i)
				} else if s, ok := v.(string); ok {
					rres[k.(string)] = s
				}
			}

			res = append(res, rres)
		}
	}

	return res, nil
}

func genAeroData(client *aero.Client, ns, set string, data []map[string]interface{}) error {
	err := client.Truncate(nil, ns, set, nil)
	if err != nil {
		return err
	}

	for i := range data {
		key, _ := aero.NewKey(ns, set, data[i]["id"])
		if err := client.Put(nil, key, data[i]); err != nil {
			return err
		}

		if i > 0 && i%1000 == 0 {
			log.Println("Aerospike progress:", i)
		}
	}

	log.Println("Aerospike completed successfully...", len(data))

	return nil
}

func aeroSetupDB(client *aero.Client, currentPath string) error {
	fileName := filepath.Join(currentPath, "aggAPI.lua")
	luaFile, err := ioutil.ReadFile(fileName)
	if err != nil {
		return err
	}

	regTask, err := client.RegisterUDF(nil, luaFile, "aggAPI.lua", aero.LUA)
	if err != nil {
		return err
	}

	// wait until UDF is created
	err = <-regTask.OnComplete()
	if err != nil {
		return err
	}

	return nil
}


test/sqlite_test.go
package main_test

import (
	"fmt"
	"log"

	"github.com/jmoiron/sqlx"
	_ "github.com/mattn/go-sqlite3"
)

// Init initializes the database backed by the file located at path.
func sqlite3db(path string) (*sqlx.DB, error) {
	log.Printf("Sqlite3 database path is %s", path)

	db, err := sqlx.Open("sqlite3", path)
	if err != nil {
		return nil, fmt.Errorf("Error connecting to the sqlite3 database: %w", err)
	}

	return db, nil
}

func genSqlData(db *sqlx.DB, data []map[string]interface{}) error {
	db.MustExec("DROP TABLE IF EXISTS test;")

	const schema = `
CREATE TABLE IF NOT EXISTS test (
    id INTEGER PRIMARY KEY,
    name TEXT NOT NULL,
    lastname TEXT NOT NULL,
    age INTEGER NOT NULL,
    salary INTEGER NOT NULL
);
`
	db.MustExec(schema)

	tx := db.MustBegin()
	for i := range data {
		_, err := tx.NamedExec("INSERT INTO test (id, name, lastname, age, salary) VALUES (:id, :name, :lastname, :age, :salary)", data[i])
		if err != nil {
			return err
		}

		if i > 0 && i%1000 == 0 {
			log.Println("sqlite3 progess:", i)
		}
	}
	tx.Commit()

	log.Println("sqlite3 completed successfully...", len(data))

	return nil
}

func sqlQuery(db *sqlx.DB, qry string) ([]map[string]interface{}, error) {
	rows, err := db.Queryx(qry)
	if err != nil {
		return nil, err
	}

	res := []map[string]interface{}{}
	for rows.Next() {
		m := map[string]interface{}{}
		err := rows.MapScan(m)
		if err != nil {
			return nil, err
		}

		res = append(res, m)
	}

	return res, nil
}


java/src/com/aerospike/examples/Main.java
/*
 * Copyright 2012-2019 Aerospike, Inc.
 *
 * Portions may be licensed to Aerospike, Inc. under one or more contributor
 * license agreements WHICH ARE COMPATIBLE WITH THE APACHE LICENSE, VERSION 2.0.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.aerospike.examples;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Arrays;

import javax.swing.JPanel;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.PosixParser;

import com.aerospike.client.Log;
import com.aerospike.client.Log.Level;
import com.aerospike.client.async.EventLoopType;
import com.aerospike.client.policy.AuthMode;
import com.aerospike.client.policy.TlsPolicy;
import com.aerospike.client.util.Util;

public class Main extends JPanel {

	private static final long serialVersionUID = 1L;
	private static final String[] ExampleNames = new String[] {
		"QueryLuaAgg"
	};
	public static String[] getAllExampleNames() { return ExampleNames; }

	/**
	 * Main entry point.
	 */
	public static void main(String[] args) {

		try {
			Options options = new Options();
			options.addOption("h", "host", true,
					"List of seed hosts in format:\n" +
					"hostname1[:tlsname][:port1],...\n" +
					"The tlsname is only used when connecting with a secure TLS enabled server. " +
					"If the port is not specified, the default port is used.\n" +
					"IPv6 addresses must be enclosed in square brackets.\n" +
					"Default: localhost\n" +
					"Examples:\n" +
					"host1\n" +
					"host1:3000,host2:3000\n" +
					"192.168.1.10:cert1:3000,[2001::1111]:cert2:3000\n"
					);
			options.addOption("p", "port", true, "Server default port (default: 3000)");
			options.addOption("U", "user", true, "User name");
			options.addOption("P", "password", true, "Password");
			options.addOption("n", "namespace", true, "Namespace (default: test)");
			options.addOption("s", "set", true, "Set name. Use 'empty' for empty set (default: demoset)");
			options.addOption("tls", "tlsEnable", false, "Use TLS/SSL sockets");
			options.addOption("tp", "tlsProtocols", true,
					"Allow TLS protocols\n" +
					"Values:  TLSv1,TLSv1.1,TLSv1.2 separated by comma\n" +
					"Default: TLSv1.2"
					);
			options.addOption("tlsCiphers", "tlsCipherSuite", true,
					"Allow TLS cipher suites\n" +
					"Values:  cipher names defined by JVM separated by comma\n" +
					"Default: null (default cipher list provided by JVM)"
					);
			options.addOption("tr", "tlsRevoke", true,
					"Revoke certificates identified by their serial number\n" +
					"Values:  serial numbers separated by comma\n" +
					"Default: null (Do not revoke certificates)"
					);
			options.addOption("tlsLoginOnly", false, "Use TLS/SSL sockets on node login only");
			options.addOption("auth", true, "Authentication mode. Values: " + Arrays.toString(AuthMode.values()));
			options.addOption("netty", false, "Use Netty NIO event loops for async examples");
			options.addOption("nettyEpoll", false, "Use Netty epoll event loops for async examples (Linux only)");
			options.addOption("g", "gui", false, "Invoke GUI to selectively run tests.");
			options.addOption("d", "debug", false, "Run in debug mode.");
			options.addOption("u", "usage", false, "Print usage.");
			options.addOption("i", "files", true, "list of files");

			CommandLineParser parser = new PosixParser();
			CommandLine cl = parser.parse(options, args, false);

			if (args.length == 0 || cl.hasOption("u")) {
				logUsage(options);
				return;
			}
			Parameters params = parseParameters(cl);
			String[] exampleNames = cl.getArgs();

			if ((exampleNames.length == 0) && (!cl.hasOption("g"))) {
				logUsage(options);
				return;
			}

			// Check for all.
			for (String exampleName : exampleNames) {
				if (exampleName.equalsIgnoreCase("all")) {
					exampleNames = ExampleNames;
					break;
				}
			}

			if (cl.hasOption("netty")) {
				params.eventLoopType = EventLoopType.NETTY_NIO;
			}

			if (cl.hasOption("nettyEpoll")) {
				params.eventLoopType = EventLoopType.NETTY_EPOLL;
			}

			if (cl.hasOption("d")) {
				Log.setLevel(Level.DEBUG);
			}

			if (cl.hasOption("g")) {
				GuiDisplay.startGui(params);
			}
			else {
				Console console = new Console();
				runExamples(console, params, exampleNames);
			}
		}
		catch (Exception ex) {
			ex.printStackTrace();
		}
	}

	/**
	 * Write usage to console.
	 */
	private static void logUsage(Options options) {
		HelpFormatter formatter = new HelpFormatter();
		StringWriter sw = new StringWriter();
		PrintWriter pw = new PrintWriter(sw);
		String syntax = Main.class.getName() + " [<options>] all|(<example1> <example2> ...)";
		formatter.printHelp(pw, 100, syntax, "options:", options, 0, 2, null);
		System.out.println(sw.toString());
		System.out.println("examples:");

		for (String name : ExampleNames) {
			System.out.println(name.toString());
		}
		System.out.println();
		System.out.println("All examples will be run if 'all' is specified as an example.");
	}

	/**
	 * Parse command line parameters.
	 */
	private static Parameters parseParameters(CommandLine cl) throws Exception {
		String host = cl.getOptionValue("h", "127.0.0.1");
		String portString = cl.getOptionValue("p", "3000");
		int port = Integer.parseInt(portString);
		String namespace = cl.getOptionValue("n","test");
		String set = cl.getOptionValue("s", "demoset");

		if (set.equals("empty")) {
			set = "";
		}

		String user = cl.getOptionValue("U");
		String password = cl.getOptionValue("P");

		if (user != null && password == null) {
			java.io.Console console = System.console();

			if (console != null) {
				char[] pass = console.readPassword("Enter password:");

				if (pass != null) {
					password = new String(pass);
				}
			}
		}

		TlsPolicy tlsPolicy = null;

		if (cl.hasOption("tls")) {
			tlsPolicy = new TlsPolicy();

			if (cl.hasOption("tp")) {
				String s = cl.getOptionValue("tp", "");
				tlsPolicy.protocols = s.split(",");
			}

			if (cl.hasOption("tlsCiphers")) {
				String s = cl.getOptionValue("tlsCiphers", "");
				tlsPolicy.ciphers = s.split(",");
			}

			if (cl.hasOption("tr")) {
				String s = cl.getOptionValue("tr", "");
				tlsPolicy.revokeCertificates = Util.toBigIntegerArray(s);
			}

			if (cl.hasOption("tlsLoginOnly")) {
				tlsPolicy.forLoginOnly = true;
			}
		}

		String fileName = cl.getOptionValue("i", "");

		AuthMode authMode = AuthMode.INTERNAL;

		if (cl.hasOption("auth")) {
			authMode = AuthMode.valueOf(cl.getOptionValue("auth", "").toUpperCase());
		}

		return new Parameters(tlsPolicy, host, port, user, password, authMode, namespace, set, fileName);
	}

	/**
	 * Connect and run one or more client examples.
	 */
	public static void runExamples(Console console, Parameters params, String[] examples) throws Exception {
		ArrayList<String> syncExamples = new ArrayList<String>();
		ArrayList<String> asyncExamples = new ArrayList<String>();

		for (String example : examples) {
			if (example.startsWith("Async")) {
				asyncExamples.add(example);
			}
			else {
				syncExamples.add(example);
			}
		}

		if (syncExamples.size() > 0) {
			Example.runExamples(console, params, syncExamples);
		}

		if (asyncExamples.size() > 0) {
			AsyncExample.runExamples(console, params, asyncExamples);
		}
	}
}


java/src/com/aerospike/examples/Console.java
/*
 * Copyright 2012-2019 Aerospike, Inc.
 *
 * Portions may be licensed to Aerospike, Inc. under one or more contributor
 * license agreements WHICH ARE COMPATIBLE WITH THE APACHE LICENSE, VERSION 2.0.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.aerospike.examples;

import java.text.SimpleDateFormat;
import java.util.Calendar;

import com.aerospike.client.Log;
import com.aerospike.client.Log.Level;

public class Console implements Log.Callback {

	private static final SimpleDateFormat Formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss z");

	public Console() {
		Log.setCallback(this);
	}

	public void info(String format, Object... args) {
		write(Level.INFO, format, args);
	}

	public void info(String message) {
		write(Level.INFO, message);
	}

	public void warn(String format, Object... args) {
		write(Level.WARN, format, args);
	}

	public void warn(String message) {
		write(Level.WARN, message);
	}

	public void error(String format, Object... args) {
		write(Level.ERROR, format, args);
	}

	public void error(String message) {
		write(Level.ERROR, message);
	}

	public void write(Level level, String format, Object... args) {
		write(level, String.format(format, args));
	}

	public void write(Level level, String message) {
		write(Formatter.format(Calendar.getInstance().getTime()) + ' ' + level + ' ' + message);
	}

	public void write(String format, Object... args) {
		write(String.format(format, args));
	}

	public void write(String message) {
		System.out.println(message);
	}

	@Override
	public void log(Level level, String message) {
		write(level, message);
	}

    public void info(int size) {
    }
}


java/src/com/aerospike/examples/Parameters.java
/*
 * Copyright 2012-2019 Aerospike, Inc.
 *
 * Portions may be licensed to Aerospike, Inc. under one or more contributor
 * license agreements WHICH ARE COMPATIBLE WITH THE APACHE LICENSE, VERSION 2.0.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.aerospike.examples;

import java.util.Map;

import com.aerospike.client.AerospikeClient;
import com.aerospike.client.Info;
import com.aerospike.client.async.EventLoopType;
import com.aerospike.client.cluster.Node;
import com.aerospike.client.policy.AuthMode;
import com.aerospike.client.policy.Policy;
import com.aerospike.client.policy.TlsPolicy;
import com.aerospike.client.policy.WritePolicy;

/**
 * Configuration data.
 */
public class Parameters {
	String host;
	int port;
	String user;
	String password;
	String namespace;
	String set;
	AuthMode authMode;
	TlsPolicy tlsPolicy;
	WritePolicy writePolicy;
	Policy policy;
	EventLoopType eventLoopType = EventLoopType.DIRECT_NIO;
	int maxCommandsInProcess;
	int maxCommandsInQueue;
	boolean singleBin;
	boolean hasGeo;
	boolean hasUdf;
	boolean hasCDTList;
	boolean hasCDTMap;
	String fileName;

	protected Parameters(TlsPolicy policy, String host, int port, String user, String password, AuthMode authMode, String namespace, String set, String fileName) {
		this.host = host;
		this.port = port;
		this.user = user;
		this.password = password;
		this.authMode = authMode;
		this.namespace = namespace;
		this.set = set;
		this.tlsPolicy = policy;
		this.fileName = fileName;
	}

	/**
	 * Some database calls need to know how the server is configured.
	 */
	protected void setServerSpecific(AerospikeClient client) throws Exception {
		Node node = client.getNodes()[0];
		String featuresFilter = "features";
		String namespaceFilter = "namespace/" + namespace;
		Map<String,String> tokens = Info.request(null, node, featuresFilter, namespaceFilter);

		String features = tokens.get(featuresFilter);
		hasGeo = false;
		hasUdf = false;
		hasCDTList = false;
		hasCDTMap = false;

		if (features != null) {
			String[] list = features.split(";");

			for (String s : list) {
				if (s.equals("geo")) {
					hasGeo = true;
				}
				else if (s.equals("udf")) {
					hasUdf = true;
				}
				else if (s.equals("cdt-list")) {
					hasCDTList = true;
				}
				else if (s.equals("cdt-map")) {
					hasCDTMap = true;
				}
			}
		}

		String namespaceTokens = tokens.get(namespaceFilter);

		if (namespaceTokens == null) {
			throw new Exception(String.format(
				"Failed to get namespace info: host=%s port=%d namespace=%s",
				host, port, namespace));
		}

		singleBin = parseBoolean(namespaceTokens, "single-bin");
	}

	private static boolean parseBoolean(String namespaceTokens, String name) {
		String search = name + '=';
		int begin = namespaceTokens.indexOf(search);

		if (begin < 0) {
			return false;
		}

		begin += search.length();
		int end = namespaceTokens.indexOf(';', begin);

		if (end < 0) {
			end = namespaceTokens.length();
		}

		String value = namespaceTokens.substring(begin, end);
		return Boolean.parseBoolean(value);
	}

	@Override
	public String toString() {
		return "Parameters: host=" + host +
				" port=" + port +
				" ns=" + namespace +
				" set=" + set +
				" single-bin=" + singleBin;
	}

	public String getBinName(String name)
	{
		// Single bin servers don't need a bin name.
		return singleBin ? "" : name;
	}
}


go/main.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"runtime"

	aero "github.com/aerospike/aerospike-client-go"
)

var (
	host        = flag.String("h", "127.0.0.1", "host")
	port        = flag.Int("p", 3000, "port")
	user        = flag.String("U", "", "User.")
	password    = flag.String("P", "", "Password.")
	currentPath = flag.String("dir", "", "Lua Path")

	// currentPath string
)

func main() {
	flag.Parse()

	runtime.GOMAXPROCS(runtime.NumCPU())

	log.SetOutput(os.Stdout)
	log.SetFlags(0)

	clientPolicy := aero.NewClientPolicy()
	if *user != "" {
		clientPolicy.User = *user
		clientPolicy.Password = *password
	}

	client, err := aero.NewClientWithPolicyAndHost(clientPolicy, aero.NewHost(*host, *port))
	if err != nil {
		log.Fatalln("Error connecting to the DB:", err)
	}
	defer client.Close()

	if len(*currentPath) == 0 {
		*currentPath, err = os.Getwd()
		if err != nil {
			log.Println(err)
		}
	}

	if err := setupDB(client); err != nil {
		log.Fatalln("Error registering the UDF:", err)
	}

	aero.SetLuaPath(filepath.Clean(*currentPath) + "/")
	if err := queryAggregate(client, "test", "test"); err != nil {
		log.Fatalln(err)
	}
}

func queryAggregate(client *aero.Client, nsName, setName string) error {
	stm := aero.NewStatement(nsName, setName)

	functionArgsMap := map[string]interface{}{
		"fields": map[string]interface{}{
			"name":              "name",
			"doesnt_exist":      "doesnt_exist",
			"max(doesnt_exist)": map[string]string{"func": "max", "expr": "rec['doesnt_exist']"},
			"max(age)":          map[string]string{"func": "max", "expr": "rec['age']"},
			"count(age)":        map[string]string{"func": "count", "expr": "rec['age'] ~= nil and 1"},
			"min(age)":          map[string]string{"func": "min", "expr": "rec['age']"},
			"sum(age*salary)":   map[string]string{"func": "sum", "expr": "(rec['age']  or 0) * (rec['salary'] or 0)"},
			"sum(age)":          map[string]string{"func": "sum", "expr": "rec['age']"},
		},
		"filter": "rec['age'] ~= nil and rec['age'] > 25",
		"group_by_fields": []string{
			"name",
			"lastname",
		},
	}

	recordset, err := client.QueryAggregate(nil, stm, "aggAPI", "select_agg_records", aero.NewValue(functionArgsMap))
	defer recordset.Close()

	if err != nil {
		return err
	}

	for result := range recordset.Results() {
		if result.Err != nil {
			return result.Err
		}

		// pp.Println(result.Record.Bins["SUCCESS"])
		fmt.Println("DONE!", len(result.Record.Bins["SUCCESS"].(map[interface{}]interface{})))
	}

	return nil
}

func setupDB(client *aero.Client) error {
	fileName := filepath.Join(*currentPath, "aggAPI.lua")
	luaFile, err := ioutil.ReadFile(fileName)
	if err != nil {
		return err
	}

	regTask, err := client.RegisterUDF(nil, luaFile, "aggAPI.lua", aero.LUA)
	if err != nil {
		return err
	}

	// wait until UDF is created
	err = <-regTask.OnComplete()
	if err != nil {
		return err
	}

	return nil
}
